### DISCUSSION TITLE
Parity between paste features and existing codebase - Studio team

### DISCUSSION THREAD

**Discussion by ashinde-git** [twilio-labs/paste#2529] [Office hours] [answered]
**Background**: Studio team is currently migrating to Paste from our legacy framework 
**Stage of design**: Actively developing
**Research**: Our current epic for this project is [STUDIORUN-539](https://issues.corp.twilio.com/browse/STUDIORUN-539)
**Links to mockups or prototype**: [Figma](https://www.figma.com/file/guUDaZU64xmwnOFXjwbz8Q/%5BWIP%5D-Widget-Config-Panel-Improvements?node-id=175%3A150974)

**Comment by kelseycavitt** [comment, marked as answer]
Hi all! 👋 
[Here is a link to today's recording](https://twilio.slack.com/archives/CBEQHHDLP/p1657818425389009).

With your desired functionality of needing to customize the text area and dropdown, we suggest that you try using the Combobox Primitive to have more control for what you need. We also encouraging ensuring the implementation is accessible from a screenreader and keyboard perspective, allowing a user to understand that a list of options is being presented and the ability to tab through or move between the options, and also have those options “announced” by a screenreader.

Per Lee’s comment, from an a11y standpoint, it could be a little complicated, the implementation could similar to a rich text area. She is going to find and share an example of this for the group! The implementation needs to be able to convey to the users what is happening, it’s not clear that only using/typing the curly brackets will open a dropdown at this time.

Once your team gives the combobox primitive a try, please let us know here that you’re ready for feedback, or if you have more questions, we’re here for you! We’ll follow up with you in a week or so.

Thanks again for all attending office hours, it was great to meet you!

**Reply by ashinde-git** [reply]
Hello.
We were able to implement the use case of triggering the combobox dropdown based on some input, in our case the trigger is '`{{`'
Whenever we select the option from the dropdown it should appear in the textarea/input component of the combobox. Also plain text and multiple selected items can co-exist in the textarea/input.
**Combo box primitive** and some deep diving in its documentation and implementation helped us come up with a custom solution.

**New Component:**

![Screen Shot 2022-07-15 at 11 48 23 AM](https://user-images.githubusercontent.com/106102933/179302494-caabae0b-92a2-48dc-8375-784be29af8b5.png)


The next challenge is the styling part of it. We need custom highlighted text for the `{{variable}}` portions of the text input. However on further investigation it seems that there is no option of injecting a JSX element, or changing inner html, or appending custom html elements to the textarea/input components.

![Screen Shot 2022-07-15 at 1 08 32 PM](https://user-images.githubusercontent.com/106102933/179303236-0c844338-4566-47d4-af35-3f00e257064d.png)

**Legacy UI:**

![Screen Shot 2022-07-15 at 12 54 11 PM](https://user-images.githubusercontent.com/106102933/179302543-89600bc3-b37c-4ee0-88f6-6b130e392481.png)

As you can see in the legacy version we have added spans with class "`mention-variable`" to handle the highlighted portions.

![Screen Shot 2022-07-15 at 1 13 25 PM](https://user-images.githubusercontent.com/106102933/179303903-edf309b3-23de-40b4-86e0-b02aa43698fc.png)


**Are there any alternatives to implement this use-case?**

**Reply by shleewhite** [reply]
Hi @ashinde-git! 

Unfortunately, the example I mentioned in office hours isn't publicly available. But it worked like this:

The area where the user types isn't a TextArea, it is actually [a content editable div](https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/contenteditable). This makes it possible to put any HTML where the user types (this will also help with styling the inserted variables).

Outside the content editable div, there is a visually hidden combobox (you can use our <ScreenReaderOnly> component to do this). When the user types "{{", their focus moves to the input of the visually hidden combobox. The dropdown of the combobox is visible and positioned absolutely over the div.

One thing to note, the visually hidden input will need a label that lets users know what this combobox is for (something like "Variables").

Doing all this makes it so that all users are able to understand what this input does and how they can interact with it. Please let me know if you have more questions!

**Reply by ashinde-git** [reply]
Thanks @shleewhite. Content editable div seems like a good idea.
So we should try custom styled divs instead of the paste textarea/input components?

**Reply by TheSisb** [reply]
Hi @ashinde-git,

It is likely you will need custom components rather than the Paste Textarea yeah. Another option is something like this: https://codersblock.com/blog/highlight-text-inside-a-textarea/

**Reply by ashinde-git** [reply]
@TheSisb That's a neat trick. Will try this out and let you know. Thank you!

**Reply by cfahlgren1** [reply]
Hey @shleewhite we are able to style the DraftJS editor so we can still have the liquid logic and have the same interface as the Paste Text Area field. Is there anything from an accessiblity standpoint we need to make sure to add so we can adhere to paste design principles even if we can't use a paste component for this usecase? I know `htmlFor` is one of them. Thank you for the help!

**Reply by shleewhite** [reply]
Hi @cfahlgren1 - we don't have guidance for WISYWIG/rich text editors yet, but the most important thing is making sure the field is properly labelled and that the interactions you can do with a mouse are possible with only a keyboard. We have some guidance on labeling in [the Textarea documentation](https://paste.twilio.design/components/textarea#accessibility).

**Reply by kate-porter** [reply]
Hey all! [Here's the recording](https://www.loom.com/share/8857e14a1ff94dccad9270a613b264bd) from your office hours session last week (password is in the #help-design-system channel in Slack). Let us know if you have any other questions!
